From 39149dcab5eb06b6bf0e8f4d9c8480eb237f7a7e Mon Sep 17 00:00:00 2001
From: John Greene <jogreene@redhat.com>
Date: Wed, 10 Sep 2014 18:28:53 -0400
Subject: [ethernet] ixgbe: check Core Clock Disable bit

Message-id: <1410373750-11341-34-git-send-email-jogreene@redhat.com>
Patchwork-id: 93536
O-Subject: [RHEL7.1 PATCH 33/50] ixgbe: check Core Clock Disable bit
Bugzilla: 1091123
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

BZ: 1091123
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7919168
Tested: In beaker. See BZ for details

This patch corrects the stop_mac_link_on_d3 function in ixgbe_82599 by
checking the Core Clock Disable bit before stopping link.

CC: Arun Sharma <asharma@fb.com>
Reported-by: Chris Pavlas <chris.pavlas@intel.com>
Signed-off-by: Jacob Keller <jacob.e.keller@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
(cherry picked from commit f68bfdb14becbce565d72ff2e8571dbb3081db9c)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/ixgbe/ixgbe_82599.c b/drivers/net/ethernet/intel/ixgbe/ixgbe_82599.c
index 71048cf..2f6c26f 100644
--- a/drivers/net/ethernet/intel/ixgbe/ixgbe_82599.c
+++ b/drivers/net/ethernet/intel/ixgbe/ixgbe_82599.c
@@ -518,8 +518,12 @@ out:
 static void ixgbe_stop_mac_link_on_d3_82599(struct ixgbe_hw *hw)
 {
  u32 autoc2_reg;
+ u16 ee_ctrl_2 = 0;
 
- if (!hw->mng_fw_enabled && !hw->wol_enabled) {
+ hw->eeprom.ops.read(hw, IXGBE_EEPROM_CTRL_2, &ee_ctrl_2);
+
+ if (!hw->mng_fw_enabled && !hw->wol_enabled &&
+     ee_ctrl_2 & IXGBE_EEPROM_CCD_BIT) {
   autoc2_reg = IXGBE_READ_REG(hw, IXGBE_AUTOC2);
   autoc2_reg |= IXGBE_AUTOC2_LINK_DISABLE_ON_D3_MASK;
   IXGBE_WRITE_REG(hw, IXGBE_AUTOC2, autoc2_reg);
diff --git a/drivers/net/ethernet/intel/ixgbe/ixgbe_type.h b/drivers/net/ethernet/intel/ixgbe/ixgbe_type.h
index 7e5f215..03e2ba5 100644
--- a/drivers/net/ethernet/intel/ixgbe/ixgbe_type.h
+++ b/drivers/net/ethernet/intel/ixgbe/ixgbe_type.h
@@ -1788,6 +1788,9 @@ enum {
 #define IXGBE_EEPROM_RD_BUFFER_MAX_COUNT 512 /* EEPROM words # read in burst */
 #define IXGBE_EEPROM_WR_BUFFER_MAX_COUNT 256 /* EEPROM words # wr in burst */
 
+#define IXGBE_EEPROM_CTRL_2 1 /* EEPROM CTRL word 2 */
+#define IXGBE_EEPROM_CCD_BIT 2 /* EEPROM Core Clock Disable bit */
+
 #ifndef IXGBE_EEPROM_GRANT_ATTEMPTS
 #define IXGBE_EEPROM_GRANT_ATTEMPTS 1000 /* EEPROM # attempts to gain grant */
 #endif
-- 
1.7.1